### ISSUE TITLE
Invalid Date sorting causes look-ahead bias in trading strategy

### ISSUE THREAD

**Issue by tislamov** [notadamking/Bitcoin-Trader-RL#28]
@notadamking There may be a potential issue with the code. In test.py when you load csv file and sort by Date, the order of time stamp becomes invalid due to time stamp format in coinbase_hourly.csv.
For example after sort by Date:
16     2019-05-16 09-AM  BTCUSD  7978.68  ...  7974.62      624.01   5005531.36
4      2019-05-16 09-PM  BTCUSD  7668.02  ...  7915.52     1285.55  10056697.28
15     2019-05-16 10-AM  BTCUSD  7974.62  ...  7849.61     2038.99  15997694.17
3      2019-05-16 10-PM  BTCUSD  7915.52  ...  7822.55      985.91   7790640.90
14     2019-05-16 11-AM  BTCUSD  7849.61  ...  7862.64      759.09   5985568.43
2      2019-05-16 11-PM  BTCUSD  7822.55  ...  7878.96      780.10   6131121.66
25     2019-05-16 12-AM  BTCUSD  8203.32  ...  8339.96     2080.89  17253110.58
13     2019-05-16 12-PM  BTCUSD  7862.64  ...  7793.87      992.68   7788906.08

You see the problem?  AM/PM and so on

**Comment by notadamking** [comment]
Yeah, I see the problem, though I will have to look into this more as I don't believe this is the case on my setup. For example, if it were, the rendered visualization of the environment would show these wild price changes.

**Comment by tislamov** [comment]
Whats interesting is that in BitcoinTradingGraph you actually cast Date to Time column and sort by Time. So visualization maybe different, than data that it used to train/test.  Wild price changes can explain excessive returns.

**Comment by rhamnett** [comment]
https://imgur.com/a/Cn7JJGX I think he is correct in his assumption

**Comment by rhamnett** [comment]
df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d %I-%p')


try this before the date sort

**Comment by notadamking** [comment]
Wow, it looks like you're right. It will be very interesting to see how this affects the results.

**Comment by rhamnett** [comment]
> Wow, it looks like you're right. It will be very interesting to see how this affects the results.

I am currently just doing a quick initial retrain to see

**Comment by notadamking** [comment]
> > Wow, it looks like you're right. It will be very interesting to see how this affects the results.
> 
> I am currently just doing a quick initial retrain to see

I am also doing a full re-train of the Sortino agent. I will post my results. Does someone want to quickly open a PR for this?

**Comment by rhamnett** [comment]
It's not a quick PR in my mind, we need to handle different date formats gracefully for different timeframe inputs. I have a class i wrote a while ago that gracefully tries to parse dates with diff formats, ill have to dig it out.

**Comment by notadamking** [comment]
This goes along well with the DataLoader proposed in #3.

**Comment by rhamnett** [comment]
If you use telegram i've set up a group with your name (and pic) on it :)  - it'll be much easier to converse there 
 - https://t.me/bitcoin_trader_rl

**Comment by notadamking** [comment]
After training/testing on the corrected data set, my first Sortino agent has achieved a profit of 45% on the test set (~4 months). I will continue to test and try to improve these results.

**Comment by tislamov** [comment]
@notadamking did you run optimizer again?  perhaps hyper params need to be changed for updated dataset

**Comment by bing007** [comment]
Just out of interest, why do we bother sorting the data by dates? Theoretically the data provider should supply the data in correct time order already. At most we just need to reversly traverse the data rows?

**Comment by silentrob** [comment]
45% is still a great return. I wonder if you can tie the confidence into the position size to adjust the risk per trade. Or have some other algorithm to find the right position size per trade.

**Comment by rhamnett** [comment]
> 45% is still a great return. I wonder if you can tie the confidence into the position size to adjust the risk per trade. Or have some other algorithm to find the right position size per trade.

Not if it's worse than buy and hold. We are reviewing the metrics and we can post some findings as and when we get them.

In answer to the sort question, you can comment it out for now but we should support getting the data into the right order automatically if we can.

**Comment by notadamking** [comment]
> @notadamking did you run optimizer again? perhaps hyper params need to be changed for updated dataset

I did not run the optimizer again, I will do this later and post the results.

**Comment by rhamnett** [comment]
I'm also running optimizer again, it's very slow.

**Comment by archenroot** [comment]
@rhamnett - you train full 1000 trials?

**Comment by rhamnett** [comment]
@archenroot no definitely not, just a few samples for now. My data is back to 2018-01-01 hourly.

**Comment by archenroot** [comment]
@rhamnett - ok, I see, just when I had 100 trials optimized, but on those best params agent wasn't executing any kind of trade :-) If I have time today, I will test PR with optuna RDBMS custom params and run optimize.py with 64 core setup... will report how this helps/or not..

**Comment by xeroquark** [comment]
> After training/testing on the corrected data set, my first Sortino agent has achieved a profit of 45% on the test set (~4 months). I will continue to test and try to improve these results.

By using the same hyper-parameters I retrained 8 Sortino agents and ran a test on Bitcoin 1h dataset. The most gain I saw was 118%. In comparison, HODL strategy gave 108%. 

Out of curiosity, I trained 5 agents with the same hyper-parameters and ran them on Bittrex 1h dataset. While still very profitable, they all performed worse than HODL. 

Has anybody had a chance to tune the hyper-parameters?

**Comment by laneshetron** [comment]
@xeroquark I've retrained Sortino with both the old hyperparams as well as new (albeit weakly-optimized) hyper-parameters and I'm seeing similar results. I've up to 56 agents trained with the new params, and it seems to have converged right around the performance of buy-and-hold. My intuition is that it's doing only a minimal amount of actual trading.

**Comment by TheExGenesis** [comment]
How many epochs are you training for? ~20 doesn't seem to be enough

**Comment by RJ4100** [comment]
Why was the hourly dataset being used? Have anyone check the difference if tick or minute data is used?

**Comment by sph3rex** [comment]
The agent is pretty much loosing with my tests over 1min/5min data. Meanwhile anyone can use cryptocompare's API to get the data:
`pip install cryptocompare`
`
df = cryptocompare.get_historical_price_hour('BTC', curr='USD')

df = pd.DataFrame(df['Data'])
df.columns = ['Close', 'High', 'Low', 'Open', 'Date', 'Volume BTC', 'Volume USD']
df['Date'] = pd.to_datetime(df['Date'],unit='s')

df.set_index('Date')
df['Date'] = df['Date'].dt.strftime('%Y-%m-%d %H')

df = df.sort_values(['Date'])
`
ccxt is also an option here :)

**Comment by nachovoss** [comment]
![Figure_1](https://user-images.githubusercontent.com/38444147/59985703-70b63a80-9633-11e9-9c14-a8a2adb0879f.png)


made some changes to the code  and it seems  to be working fine, let me know if you face any errors, or if i made any mistakes:

CitcoinTradingGraph.py line 22     lambda x: datetime.strptime(x, '%Y-%m-%d %H:%M:%S')

DATAFRAME FOR TEST , TRAIN AND OPTIMIZE

df = pd.read_csv('./data/coinbase_hourly.csv')
df = df.drop(['Symbol'], axis=1)
df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d %I-%p')
df['Date'] = df['Date'].astype(str)
df = df.sort_values(['Date'])
df = add_indicators(df.reset_index())

**Comment by notadamking** [comment]
Fixed in #23
